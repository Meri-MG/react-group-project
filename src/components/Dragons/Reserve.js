import PropTypes from 'prop-types';

function Reserve({ reserveHandler, content, styling }) {
  return (
    <button className={styling} onClick={reserveHandler} type="button">
      {content}
    </button>
  );
}

Reserve.propTypes = {
  reserveHandler: PropTypes.func.isRequired,
  content: PropTypes.string.isRequired,
  styling: PropTypes.string.isRequired,
};
export default Reserve;
