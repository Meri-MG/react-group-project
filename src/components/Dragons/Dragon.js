import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { v4 as uuidv4 } from 'uuid';
import store from '../../redux/configureStore';
import Reserve from './Reserve';
import classes from './Dragons.module.css';
import { reserveDragon } from '../../redux/dragons/dragons';

const getDataFromApi = 'getDataFromApi';
function Dragon() {
  const dragons = useSelector((state) => state.dragonReducer);
  useEffect(
    () => {
      store.dispatch({ type: getDataFromApi });
    }, [],
  );
  const reserveHandler = (el) => {
    store.dispatch(reserveDragon(el));
  };
  return (
    <div>
      <ul className={classes.ul}>
        {dragons.map((el) => (
          <li key={el.id}>
            <span>
              {el.flickr_images.map((img) => <img key={uuidv4()} src={img} alt={el.name} />)}
            </span>
            {el.reserved ? <span className={classes.badge}>Reserved</span> : ''}
            <span>
              {el.description}
            </span>
            <span>
              {el.name}
            </span>
            <span><Reserve reserveHandler={() => reserveHandler(el)} content={el.reserved ? 'Cancel reservation' : 'Reserve dragon'} /></span>
          </li>
        ))}
      </ul>
    </div>
  );
}

export default Dragon;
