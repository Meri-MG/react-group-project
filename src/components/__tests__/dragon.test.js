import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import store from '../../redux/configureStore';
import Dragon from '../Dragons/Dragon';
import Reserve from '../Dragons/Reserve';

describe('Test Dragon section functionality ', () => {
  test('snapshot', () => {
    const tree = renderer.create(<Provider store={store}><Dragon /></Provider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test('test click ', async () => {
    const reserveHandler = jest.fn();
    const el = { reserved: false };
    render(<Reserve
      reserveHandler={
        () => reserveHandler(el)
      }
      content={el.reserved ? 'Cancel reservation' : 'Reserve dragon'}
    />);
    await userEvent.click(screen.getByRole('button'));
    expect(reserveHandler).toHaveBeenCalled();
  });
  test('test button text ', async () => {
    const reserveHandler = jest.fn();
    const el = { reserved: true };
    render(<Reserve
      reserveHandler={
        () => reserveHandler(el)
      }
      content={el.reserved ? 'Cancel reservation' : 'Reserve dragon'}
    />);
    await userEvent.click(screen.getByRole('button'));
    expect(screen.getByRole('button').textContent).toEqual('Cancel reservation');
  });
});
