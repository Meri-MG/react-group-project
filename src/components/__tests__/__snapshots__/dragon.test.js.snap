// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Dragon section functionality  snapshot 1`] = `
<div>
  <ul
    className="ul"
  />
</div>
`;
